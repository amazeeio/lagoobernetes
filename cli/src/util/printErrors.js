// @flow

import { red } from 'chalk';
import R from 'ramda';

type Cerr = typeof console.error;
type LaguError = { +message: string, +stack?: string };

export function printErrors(
  cerr: Cerr,
  ...errors: Array<LaguError | string>
): number {
  R.compose(
    R.forEach(cerr),
    R.map(red),
    R.map(
      err =>
        typeof err === 'string'
          ? err
          : R.propOr(
            // ...otherwise, use the message property as a fallback
            R.prop('message', err),
            // Try to get the stack...
            'stack',
            err,
          ),
    ),
  )(errors);
  return 1;
}

export function printNoConfigError(cerr: Cerr): number {
  return printErrors(cerr, {
    message:
      '.lagoobernetes.yml config file not found. Please create one with "lagoobernetes init".\nOnline documentation: https://github.com/amazeeio/lagoobernetes/blob/master/cli/README.md#lagoobernetes-init',
  });
}

export function printProjectConfigurationError(cerr: Cerr): number {
  return printErrors(cerr, {
    message:
      'No project configured. Please create a .lagoobernetes.yml config file with "lagoobernetes init" or pass a project to this command via the --project option.\nOnline documentation: https://github.com/amazeeio/lagoobernetes/blob/master/cli/README.md#lagoobernetes-init',
  });
}

export function printGraphQLErrors(
  cerr: Cerr,
  ...errors: Array<LaguError>
): number {
  if (
    R.find(R.propEq('message', 'Unauthorized - Bearer Token Required'))(errors)
  ) {
    return printErrors(cerr, {
      message:
        'No authentication token found. Please log in first with "lagoobernetes login".\nOnline documentation: https://github.com/amazeeio/lagoobernetes/blob/master/cli/README.md#lagoobernetes-login',
    });
  }
  const errorsPluralized = R.length(errors) === 1 ? 'an error' : 'errors';
  return printErrors(
    cerr,
    `Oops! The Lagoobernetes API returned ${errorsPluralized}:`,
    ...errors,
  );
}
