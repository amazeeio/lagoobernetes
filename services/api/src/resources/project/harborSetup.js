// @flow

/* ::
import type MariaSQL from 'mariasql';
*/

const harborClient = require('../../clients/harborClient');
const logger = require('../../logger');
const R = require('ramda');
const Sql = require('../env-variables/sql');
const { isPatchEmpty, prepare, query, whereAnd } = require('../../util/db');

const lagoobernetesHarborRoute = R.compose(
  R.defaultTo('http://172.17.0.1:8084'),
  R.find(R.test(/harbor-nginx/)),
  R.split(','),
  R.propOr('', 'LAGOOBERNETES_ROUTES'),
)(process.env);

const createHarborOperations = (sqlClient /* : MariaSQL */) => ({
  addProject: async (lagoobernetesProjectName, projectID) => {
    // Create harbor project
    try {
      var res = await harborClient.post(`projects`, {
        body: {
          count_limit: -1,
          project_name: lagoobernetesProjectName,
          storage_limit: -1,
          metadata: {
            auto_scan: "true",
            reuse_sys_cve_whitelist: "true",
            public: "false"
          }
        }
      });
      logger.debug(`Harbor project ${lagoobernetesProjectName} created!`)
    } catch (err) {
      // 409 means project already exists
      // 201 means project created successfully
      if (err.statusCode == 409) {
        logger.info(`Unable to create the harbor project "${lagoobernetesProjectName}", as it already exists in harbor!`)
      } else {
        console.log(res)
        logger.error(`Unable to create the harbor project "${lagoobernetesProjectName}", error: ${err}`)
      }
    }

    // Get new harbor project's id
    try {
      const res = await harborClient.get(`projects?name=${lagoobernetesProjectName}`)
      var harborProjectID = res.body[0].project_id
      logger.debug(`Got the harbor project id for project ${lagoobernetesProjectName} successfully!`)
    } catch (err) {
      if (err.statusCode == 404) {
        logger.error(`Unable to get the harbor project id of "${lagoobernetesProjectName}", as it does not exist in harbor!`)
      } else {
        logger.error(`Unable to get the harbor project id of "${lagoobernetesProjectName}" !!`)
      }
    }

    logger.debug(`Harbor project id for ${lagoobernetesProjectName}: ${harborProjectID}`)
    // Create robot account for new harbor project
    try {
      const res = await harborClient.post(`projects/${harborProjectID}/robots`, {
        body: {
          name: lagoobernetesProjectName,
          access: [
            {
              resource: `/project/${harborProjectID}/repository`,
              action: "push"
            }
          ]
        }
      })
      var harborTokenInfo = res.body
      logger.debug(`Robot was created for Harbor project ${lagoobernetesProjectName} !`)
    } catch (err) {
      // 409 means project already exists
      // 201 means project created successfully
      if (err.statusCode == 409) {
        logger.warn(`Unable to create a robot account for harbor project "${lagoobernetesProjectName}", as a robot account of the same name already exists!`)
      } else {
        logger.warn(`Unable to create a robot account for harbor project "${lagoobernetesProjectName}" !!`)
      }
    }

    // Set Harbor env vars for lagoobernetes environment
    try {
      await query(
        sqlClient,
        Sql.insertEnvVariable({
          "name": "INTERNAL_REGISTRY_URL",
          "value": lagoobernetesHarborRoute,
          "scope": "INTERNAL_CONTAINER_REGISTRY",
          "project": projectID,
        }),
      );
      logger.debug(`Environment variable INTERNAL_REGISTRY_URL for ${lagoobernetesProjectName} created!`)

      await query(
        sqlClient,
        Sql.insertEnvVariable({
          "name": "INTERNAL_REGISTRY_USERNAME",
          "value": harborTokenInfo.name,
          "scope": "INTERNAL_CONTAINER_REGISTRY",
          "project": projectID,
        }),
      );
      logger.debug(`Environment variable INTERNAL_REGISTRY_USERNAME for ${lagoobernetesProjectName} created!`)

      await query(
        sqlClient,
        Sql.insertEnvVariable({
          "name": "INTERNAL_REGISTRY_PASSWORD",
          "value": harborTokenInfo.token,
          "scope": "INTERNAL_CONTAINER_REGISTRY",
          "project": projectID,
        }),
      );
      logger.debug(`Environment variable INTERNAL_REGISTRY_PASSWORD for ${lagoobernetesProjectName} created!`)
    } catch (err) {
      logger.error(`Error while setting up harbor environment variables for ${lagoobernetesProjectName}, error: ${err}`)
    }
  }
})

module.exports = createHarborOperations;
