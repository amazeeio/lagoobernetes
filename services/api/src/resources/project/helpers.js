// @flow

/* ::
import type MariaSQL from 'mariasql';
*/

const R = require('ramda');
const { asyncPipe } = require('@lagoobernetes/commons/src/util');
const { query } = require('../../util/db');
const logger = require('../../logger');

const Sql = require('./sql');

const Helpers = (sqlClient /* : MariaSQL */) => {
  const getProjectById = async (id /* : string */) => {
    const rows = await query(sqlClient, Sql.selectProject(id));
    return R.prop(0, rows);
  };

  const getProjectByName = async (name /* : string */) => {
    const rows = await query(sqlClient, Sql.selectProjectByName(name));
    return R.prop(0, rows);
  };

  const getProjectByEnvironmentId = async (environmentId) => {
    const rows = await query(sqlClient, Sql.selectProjectByEnvironmentId(environmentId));
    return R.prop(0, rows);
  }

  const getProjectsByIds = (projectIds /* : Array<number> */) =>
    query(sqlClient, Sql.selectProjectsByIds(projectIds));

  const getProjectsWithoutDirectUserAccess = async (
    projectIds /* : Array<string> */,
    userIds /* : Array<number> */,
  ) =>
    query(
      sqlClient,
      Sql.selectProjectsWithoutDirectUserAccess(projectIds, userIds),
    );

  return {
    getProjectById,
    getProjectByName,
    getProjectsByIds,
    getProjectByEnvironmentId,
    getProjectsWithoutDirectUserAccess,
    getProjectIdByName: async (name /* : string */) => {
      const pidResult = await query(sqlClient, Sql.selectProjectIdByName(name));

      const amount = R.length(pidResult);
      if (amount > 1) {
        throw new Error(
          `Multiple project candidates for '${name}' (${amount} found). Do nothing.`,
        );
      }

      if (amount === 0) {
        throw new Error(`Not found: '${name}'`);
      }

      const pid = R.path(['0', 'id'], pidResult);

      return pid;
    },
    getProjectByProjectInput: async projectInput => {
      const notEmpty = R.complement(R.anyPass([R.isNil, R.isEmpty]));
      const hasId = R.both(R.has('id'), R.propSatisfies(notEmpty, 'id'));
      const hasName = R.both(R.has('name'), R.propSatisfies(notEmpty, 'name'));

      const projectFromId = asyncPipe(R.prop('id'), getProjectById, project => {
        if (!project) {
          throw new Error('Unauthorized');
        }

        return project;
      });

      const projectFromName = asyncPipe(R.prop('name'), async name => {
        const rows = await query(sqlClient, Sql.selectProjectByName(name));
        const project = R.prop(0, rows);

        if (!project) {
          throw new Error('Unauthorized');
        }

        return project;
      });

      return R.cond([
        [hasId, projectFromId],
        [hasName, projectFromName],
        [
          R.T,
          () => {
            throw new Error('Must provide project "id" or "name"');
          },
        ],
      ])(projectInput);
    },
    getAllProjects: async () => query(sqlClient, Sql.selectAllProjects()),
    getAllProjectsNotIn: async ids =>
      query(sqlClient, Sql.selectAllProjectNotIn(ids)),
    getCustomerProjectsWithoutDirectUserAccess: async (
      customerIds /* : Array<number> */,
      userIds /* : Array<number> */,
    ) =>
      query(
        sqlClient,
        Sql.selectCustomerProjectsWithoutDirectUserAccess(customerIds, userIds),
      ),
    getAllProjectNames: async () =>
      R.map(
        R.prop('name'),
        await query(sqlClient, Sql.selectAllProjectNames()),
      ),
  };
};

module.exports = Helpers;
