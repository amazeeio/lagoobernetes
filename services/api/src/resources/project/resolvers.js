// @flow

const R = require('ramda');
const validator = require('validator');
const sshpk = require('sshpk');
const opendistroSecurityClient = require('../../clients/opendistroSecurityClient');
const logger = require('../../logger');
const {
  inClause,
  prepare,
  query,
  whereAnd,
  isPatchEmpty,
} = require('../../util/db');

const Helpers = require('./helpers');
const KeycloakOperations = require('./keycloak');
const { OpendistroSecurityOperations } = require('../group/opendistroSecurity');
const Sql = require('./sql');
const { generatePrivateKey, getSshKeyFingerprint } = require('../sshKey');
const sshKeySql = require('../sshKey/sql');
const harborClient = require('../../clients/harborClient');
const createHarborOperations = require('./harborSetup');

/* ::

import type {ResolversObj} from '../';

*/

const removePrivateKey = R.assoc('privateKey', null);

const isValidGitUrl = value =>
  /(?:git|ssh|https?|git@[-\w.]+):(\/\/)?(.*?)(\.git)(\/?|\#[-\d\w._]+?)$/.test(value);

const getAllProjects = async (
  root,
  args,
  {
    sqlClient,
    hasPermission,
    models,
    keycloakGrant,
  },
) => {
  let where;
  try {
    await hasPermission('project', 'viewAll');

    where = whereAnd([
      args.createdAfter ? 'created >= :created_after' : '',
      args.gitUrl ? 'git_url = :git_url' : '',
    ]);
  } catch (err) {
    if (!keycloakGrant) {
      logger.warn('No grant available for getAllProjects');
      return [];
    }

    const userProjectIds = await models.UserModel.getAllProjectsIdsForUser({
      id: keycloakGrant.access_token.content.sub,
    });

    where = whereAnd([
      args.createdAfter ? 'created >= :created_after' : '',
      args.gitUrl ? 'git_url = :git_url' : '',
      inClause('id', userProjectIds),
    ]);
  }

  const order = args.order ? ` ORDER BY ${R.toLower(args.order)} ASC` : '';

  const prep = prepare(sqlClient, `SELECT * FROM project ${where}${order}`);
  const rows = await query(sqlClient, prep(args));

  // This resolver is used for the main UI page and is quite slow. Since we've
  // already authorized the user has access to all the projects we are
  // returning, AND all user roles are allowed to view all environments, we can
  // short-circuit the slow keycloak check in the getEnvironmentsByProjectId
  // resolver.
  //
  // @TODO: When this performance issue is fixed for real, remove this hack as
  // it hardcodes a "everyone can view environments" authz rule.
  return rows.map(row => ({ ...row, environmentAuthz: true }));
};

const getProjectByEnvironmentId = async (
  { id: eid },
  args,
  {
    sqlClient,
    hasPermission,
  },
) => {
  const prep = prepare(
    sqlClient,
    `SELECT
        p.*
      FROM environment e
      JOIN project p ON e.project = p.id
      WHERE e.id = :eid
      LIMIT 1
    `,
  );

  const rows = await query(sqlClient, prep({ eid }));

  const project = rows[0];

  await hasPermission('project', 'view', {
    project: project.id,
  });

  try {
    await hasPermission('project', 'viewPrivateKey', {
      project: project.id,
    });

    return project;
  } catch (err) {
    return removePrivateKey(project);
  }
};

const getProjectByGitUrl = async (
  root,
  args,
  {
    sqlClient,
    hasPermission,
  },
) => {
  const str = `
      SELECT
        *
      FROM project
      WHERE git_url = :git_url
      LIMIT 1
    `;

  const prep = prepare(sqlClient, str);
  const rows = await query(sqlClient, prep(args));

  const project = rows[0];

  await hasPermission('project', 'view', {
    project: project.id,
  });

  try {
    await hasPermission('project', 'viewPrivateKey', {
      project: project.id,
    });

    return project;
  } catch (err) {
    return removePrivateKey(project);
  }
};

const getProjectByName = async (
  root,
  args,
  {
    sqlClient,
    hasPermission,
  },
) => {
  const str = `
      SELECT
        *
      FROM project
      WHERE name = :name
    `;

  const prep = prepare(sqlClient, str);

  const rows = await query(sqlClient, prep(args));
  const project = rows[0];

  await hasPermission('project', 'view', {
    project: project.id,
  });

  try {
    await hasPermission('project', 'viewPrivateKey', {
      project: project.id,
    });

    return project;
  } catch (err) {
    return removePrivateKey(project);
  }
};

const addProject = async (
  root,
  { input },
  {
    hasPermission,
    sqlClient,
    models,
  },
) => {
  await hasPermission('project', 'add');

  if (validator.matches(input.name, /[^0-9a-z-]/)) {
    throw new Error(
      'Only lowercase characters, numbers and dashes allowed for name!',
    );
  }
  if (!isValidGitUrl(input.gitUrl)) {
    throw new Error('The provided gitUrl is invalid.',);
  }

  let keyPair = {};
  try {
    const privateKey = R.cond([
      [R.isNil, generatePrivateKey],
      [R.isEmpty, generatePrivateKey],
      [R.T, sshpk.parsePrivateKey],
    ])(R.prop('privateKey', input));

    const publicKey = privateKey.toPublic();

    keyPair = {
      ...keyPair,
      private: R.replace(/\n/g, '\n', privateKey.toString('openssh')),
      public: publicKey.toString(),
    };
  } catch (err) {
    throw new Error(`There was an error with the privateKey: ${err.message}`);
  }

  const prep = prepare(
    sqlClient,
    `CALL CreateProject(
        :id,
        :name,
        :git_url,
        ${input.availability ? ':availability' : '"STANDARD"'},
        :private_key,
        ${input.subfolder ? ':subfolder' : 'NULL'},
        :openshift,
        ${
  input.openshiftProjectPattern ? ':openshift_project_pattern' : 'NULL'
},
        ${
  input.activeSystemsDeploy
    ? ':active_systems_deploy'
    : '"lagoobernetes_openshiftBuildDeploy"'
},
        ${
  input.activeSystemsPromote
    ? ':active_systems_promote'
    : '"lagoobernetes_openshiftBuildDeploy"'
},
        ${
  input.activeSystemsRemove
    ? ':active_systems_remove'
    : '"lagoobernetes_openshiftRemove"'
},
        ${
  input.activeSystemsTask
    ? ':active_systems_task'
    : '"lagoobernetes_openshiftJob"'
},
        ${
  input.activeSystemsMisc
    ? ':active_systems_misc'
    : '"lagoobernetes_openshiftMisc"'
},
        ${input.branches ? ':branches' : '"true"'},
        ${input.pullrequests ? ':pullrequests' : '"true"'},
        ${input.productionEnvironment ? ':production_environment' : 'NULL'},
        ${input.autoIdle ? ':auto_idle' : '1'},
        ${input.storageCalc ? ':storage_calc' : '1'},
        ${
  input.developmentEnvironmentsLimit
    ? ':development_environments_limit'
    : '5'
}
      );
    `,
  );

  const rows = await query(sqlClient, prep({
    ...input,
    privateKey: keyPair.private,
  }));
  const project = R.path([0, 0], rows);

  // Create a default group for this project
  let group;
  try {
    group = await models.GroupModel.addGroup({
      name: `project-${project.name}`,
      attributes: {
        type: ['project-default-group'],
        'lagoobernetes-projects': [project.id],
      },
    });
  } catch (err) {
    logger.error(`Could not create default project group for ${project.name}: ${err.message}`);
  }

  OpendistroSecurityOperations(sqlClient, models.GroupModel).syncGroup(`project-${project.name}`, project.id);

  // Find or create a user that has the public key linked to them
  const userRows = await query(
    sqlClient,
    sshKeySql.selectUserIdsBySshKeyFingerprint(getSshKeyFingerprint(keyPair.public)),
  );
  const userId = R.path([0, 'usid'], userRows);

  let user;
  if (!userId) {
    try {
      user = await models.UserModel.addUser({
        email: `default-user@${project.name}`,
        username: `default-user@${project.name}`,
        comment: `autogenerated user for project ${project.name}`,
      });

      const keyParts = keyPair.public.split(' ');

      const {
        info: { insertId },
      } = await query(
        sqlClient,
        sshKeySql.insertSshKey({
          id: null,
          name: 'auto-add via api',
          keyValue: keyParts[1],
          keyType: keyParts[0],
          keyFingerprint: getSshKeyFingerprint(keyPair.public),
        }),
      );
      await query(sqlClient, sshKeySql.addSshKeyToUser({ sshKeyId: insertId, userId: user.id }));
    } catch (err) {
      logger.error(`Could not create default project user for ${project.name}: ${err.message}`);
    }
  } else {
    user = await models.UserModel.loadUserById(userId);
  }

  // Add the user (with linked public key) to the default group with maintainer role
  try {
    await models.GroupModel.addUserToGroup(user, group, 'maintainer');
  } catch (err) {
    logger.error(`Could not link user to default projet group for ${project.name}: ${err.message}`);
  }

  const harborOperations = createHarborOperations(sqlClient);

  const harborResults = await harborOperations.addProject(project.name, project.id)

  return project;
};

const deleteProject = async (
  root,
  { input: { project: projectName } },
  {
    sqlClient,
    hasPermission,
    models,
  },
) => {
  // Will throw on invalid conditions
  const pid = await Helpers(sqlClient).getProjectIdByName(projectName);
  const project = await Helpers(sqlClient).getProjectById(pid);

  await hasPermission('project', 'delete', {
    project: pid,
  });

  const prep = prepare(sqlClient, 'CALL DeleteProject(:name)');
  await query(sqlClient, prep(project));

  // Remove the default group and user
  try {
    const group = await models.GroupModel.loadGroupByName(`project-${project.name}`);
    await models.GroupModel.deleteGroup(group.id);
    OpendistroSecurityOperations(sqlClient, models.GroupModel).deleteGroup(group.name);
  } catch (err) {
    logger.error(`Could not delete default group for project ${project.name}: ${err.message}`);
  }

  try {
    const user = await models.UserModel.loadUserByUsername(`default-user@${project.name}`);
    await models.UserModel.deleteUser(user.id);
  } catch (err) {
    logger.error(`Could not delete default user for project ${project.name}: ${err.message}`);
  }

  return 'success';
};

const updateProject = async (
  root,
  {
    input: {
      id,
      patch,
      patch: {
        name,
        gitUrl,
        availability,
        privateKey,
        subfolder,
        activeSystemsDeploy,
        activeSystemsRemove,
        activeSystemsTask,
        activeSystemsMisc,
        branches,
        productionEnvironment,
        autoIdle,
        storageCalc,
        pullrequests,
        openshift,
        openshiftProjectPattern,
        developmentEnvironmentsLimit,
      },
    },
  },
  {
    sqlClient,
    hasPermission,
    models,
  },
) => {
  await hasPermission('project', 'update', {
    project: id,
  });

  if (isPatchEmpty({ patch })) {
    throw new Error('input.patch requires at least 1 attribute');
  }

  if (typeof name === 'string') {
    if (validator.matches(name, /[^0-9a-z-]/)) {
      throw new Error(
        'Only lowercase characters, numbers and dashes allowed for name!',
      );
    }
  }

  if (gitUrl !== undefined && !isValidGitUrl(gitUrl)) {
    throw new Error('The provided gitUrl is invalid.',);
  }

  const oldProject = await Helpers(sqlClient).getProjectById(id);

  // TODO If the privateKey changes, automatically remove the old one from the
  // default user and link the new one.

  // const originalProject = await Helpers(sqlClient).getProjectById(id);
  // const originalName = R.prop('name', originalProject);
  // const originalCustomer = parseInt(R.prop('customer', originalProject));

  // // If the project will be updating the `name` or `customer` fields, update Keycloak groups and users accordingly
  // if (typeof customer === 'number' && customer !== originalCustomer) {
  //   // Delete Keycloak users from original projects where given user ids do not have other access via `project_user` (projects where the user loses access if they lose customer access).
  //   await Helpers(sqlClient).mapIfNoDirectProjectAccess(
  //     id,
  //     originalCustomer,
  //     async ({
  //       keycloakUserId,
  //       keycloakUsername,
  //       keycloakGroupId,
  //       keycloakGroupName,
  //     }) => {
  //       await keycloakAdminClient.users.delFromGroup({
  //         id: keycloakUserId,
  //         groupId: keycloakGroupId,
  //       });
  //       logger.debug(
  //         `Removed Keycloak user ${keycloakUsername} from group "${keycloakGroupName}"`,
  //       );
  //     },
  //   );
  // }

  await query(
    sqlClient,
    Sql.updateProject({
      id,
      patch: {
        name,
        gitUrl,
        availability,
        privateKey,
        subfolder,
        activeSystemsDeploy,
        activeSystemsRemove,
        activeSystemsTask,
        activeSystemsMisc,
        branches,
        productionEnvironment,
        autoIdle,
        storageCalc,
        pullrequests,
        openshift,
        openshiftProjectPattern,
        developmentEnvironmentsLimit,
      },
    }),
  );

  // Rename the default group and user
  if (patch.name && oldProject.name !== patch.name) {
    try {
      const group = await models.GroupModel.loadGroupByName(`project-${oldProject.name}`);
      await models.GroupModel.updateGroup({
        id: group.id,
        name: `project-${patch.name}`,
      });
    } catch (err) {
      logger.error(`Could not rename default group for project ${patch.name}: ${err.message}`);
    }

    try {
      const user = await models.UserModel.loadUserByUsername(`default-user@${oldProject.name}`);
      await models.UserModel.updateUser({
        id: user.id,
        email: `default-user@${patch.name}`,
        username: `default-user@${patch.name}`,
        comment: `autogenerated user for project ${patch.name}`,
      });
    } catch (err) {
      logger.error(`Could not rename default user for project ${patch.name}: ${err.message}`);
    }
  }

  // if (typeof name === 'string' && name !== originalName) {
  //   const groupId = await KeycloakOperations.findGroupIdByName(originalName);

  //   await keycloakAdminClient.groups.update({ id: groupId }, { name });
  //   logger.debug(
  //     `Renamed Keycloak group ${groupId} from "${originalName}" to "${name}"`,
  //   );
  // }

  // if (typeof customer === 'number' && customer !== originalCustomer) {
  //   // Add Keycloak users to new projects where given user ids do not have other access via `project_user` (projects where the user loses access if they lose customer access).
  //   await Helpers(sqlClient).mapIfNoDirectProjectAccess(
  //     id,
  //     customer,
  //     async ({
  //       keycloakUserId,
  //       keycloakUsername,
  //       keycloakGroupId,
  //       keycloakGroupName,
  //     }) => {
  //       await keycloakAdminClient.users.addToGroup({
  //         id: keycloakUserId,
  //         groupId: keycloakGroupId,
  //       });
  //       logger.debug(
  //         `Added Keycloak user ${keycloakUsername} to group "${keycloakGroupName}"`,
  //       );
  //     },
  //   );
  // }

  return Helpers(sqlClient).getProjectById(id);
};


const deleteAllProjects = async (
  root,
  args,
  { sqlClient, hasPermission },
) => {
  await hasPermission('project', 'deleteAll');

  const projectNames = await Helpers(sqlClient).getAllProjectNames();

  await query(sqlClient, Sql.truncateProject());

  for (const name of projectNames) {
    await KeycloakOperations.deleteGroup(name);
  }

  // TODO: Check rows for success
  return 'success';
};

const Resolvers /* : ResolversObj */ = {
  deleteProject,
  addProject,
  getProjectByName,
  getProjectByGitUrl,
  getProjectByEnvironmentId,
  getAllProjects,
  updateProject,
  deleteAllProjects,
};

module.exports = Resolvers;
