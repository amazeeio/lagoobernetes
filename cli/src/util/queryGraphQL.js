// @flow

import R from 'ramda';
import { getConfig } from '../config';
import { getApiConfig } from '../config/getApiConfig';
import { fileExists, readFile } from './fs';
import request from './request';
import { printErrors } from './printErrors';

type QLQueryArgs = {
  cerr: typeof console.error,
  query: string,
  variables?: Object,
  headers?: Object,
  pretty?: boolean,
};

export async function queryGraphQL({
  cerr,
  query,
  variables,
  headers: customHeaders = {},
  pretty = false,
}:
QLQueryArgs): Object {
  const headers = {
    Accept: 'application/json',
    'Content-Type': 'application/json',
    ...customHeaders,
  };

  if (!headers.Authorization) {
    const tokenFile = R.prop('token', getConfig());
    const tokenFileExists = await fileExists(tokenFile);

    if (tokenFileExists) {
      const tokenBuffer = await readFile(tokenFile);
      const token = tokenBuffer.toString().replace(/(\r\n|\n|\r)/gm, '');
      headers.Authorization = `Bearer ${token}`;
    }
  }

  const { hostname, port } = getApiConfig();

  const protocol = R.equals(port, 443) ? 'https:' : 'http:';

  const options = {
    hostname,
    path: '/graphql',
    port,
    protocol,
    method: 'POST',
    headers,
    body: JSON.stringify(
      {
        query,
        variables,
      },
      null,
      pretty ? 2 : 0,
    ),
    rejectUnauthorized: false,
  };

  let response;

  try {
    response = await request(options);
  } catch (err) {
    const error = R.ifElse(
      // For socket hang ups...
      R.propEq('message', 'socket hang up'),
      // ...print a nicer error message...
      R.always(
        `Could not connect to Lagoobernetes API at ${protocol}//${hostname}:${port}/graphql.`,
      ),
      // ...otherwise just return the error message
      R.prop('message'),
    )(err);
    printErrors(cerr, error);
    process.exit(1);
  }

  return response;
}
