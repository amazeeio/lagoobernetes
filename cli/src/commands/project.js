// @flow

import inquirer from 'inquirer';
import R from 'ramda';
import { answerWithOptionIfSetOrPrompt } from '../cli/answerWithOption';
import { setConfigForHandlers } from '../cli/setConfigForHandlers';
import format from '../util/format';
import gql from '../util/gql';
import { queryGraphQL } from '../util/queryGraphQL';
import { printGraphQLErrors } from '../util/printErrors';

import typeof Yargs from 'yargs';
import type { CommandHandlerArgsWithOptions } from '../types/Command';

export const command = 'project';
export const description = 'Show project details';

export const PROJECT: 'project' = 'project';

export const commandOptions = {
  [PROJECT]: PROJECT,
};

type OptionalOptions = {
  project?: string,
};

type Options = {
  +project: string,
};

export function builder(yargs: Yargs): Yargs {
  return yargs
    .usage(`$0 ${command} - ${description}`)
    .options({
      [PROJECT]: {
        demandOption: false,
        describe: 'Name of project',
        type: 'string',
      },
    })
    .example(
      `$0 ${command}`,
      'Show details for the project configured in .lagoobernetes.yml',
    )
    .example(
      `$0 ${command} --${PROJECT} myproject`,
      'Show details of project "myproject"',
    )
    .commandDir('projectCommands', { visit: setConfigForHandlers });
}

type PromptForQueryOptionsArgs = {|
  options: OptionalOptions,
  clog: typeof console.log,
|};

async function promptForQueryOptions({
  options,
  clog,
}:
PromptForQueryOptionsArgs): Promise<Options> {
  return inquirer.prompt([
    {
      type: 'input',
      name: PROJECT,
      message: 'Project name:',
      when: answerWithOptionIfSetOrPrompt({ option: PROJECT, options, clog }),
    },
  ]);
}

type Args = CommandHandlerArgsWithOptions<{
  +project?: string,
}>;

export async function handler({ clog, cerr, options }: Args): Promise<number> {
  const { project: projectName } = await promptForQueryOptions({
    options,
    clog,
  });

  const result = await queryGraphQL({
    cerr,
    query: gql`
      query ProjectByName($project: String!) {
        projectByName(name: $project) {
          name
          customer {
            name
          }
          gitUrl
          activeSystemsDeploy
          activeSystemsRemove
          branches
          pullrequests
          openshift {
            name
          }
          created
        }
      }
    `,
    variables: { project: projectName },
  });

  const { errors } = result;
  if (errors != null) {
    return printGraphQLErrors(cerr, ...errors);
  }

  const project = R.path(['data', 'projectByName'])(result);

  if (project == null) {
    clog(`No project '${projectName}' found.`);
    return 0;
  }

  clog(
    format([
      [
        'Name',
        'Customer',
        'Git URL',
        'Active Systems Deploy',
        'Active Systems Remove',
        'Branches',
        'Pull Requests',
        'Openshift',
        'Created',
      ],
      [
        R.prop('name', project),
        R.path(['customer', 'name'], project),
        R.prop('gitUrl', project),
        R.prop('activeSystemsDeploy', project),
        R.prop('activeSystemsRemove', project),
        R.prop('branches', project),
        R.prop('pullrequests', project),
        R.path(['openshift', 'name'], project),
        R.path(['created'], project),
      ],
    ]),
  );

  return 0;
}
