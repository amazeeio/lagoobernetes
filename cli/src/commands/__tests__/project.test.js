// @flow

import { queryGraphQL } from '../../util/queryGraphQL';
import { handler } from '../project';

jest.mock('../../util/queryGraphQL');
jest.mock('../../config', () => ({
  getConfig: jest.fn(() => ({ format: 'table' })),
}));

// Flow does not know which objects are actual mocks
// this function casts given parameter to JestMockFn
const _castMockForFlow = (mockFn: any): JestMockFn<any, any> => mockFn;

const mockErrorResponse = {
  errors: [{ message: 'something something error' }],
};

describe('handler', () => {
  const mockResponse1 = {
    data: {
      projectByName: {
        name: 'credentialstest-project1',
        customer: {
          name: 'credentialtest-customer1',
        },
        gitUrl: 'project1.git',
        activeSystemsDeploy: 'lagoobernetes_openshiftBuildDeploy',
        activeSystemsRemove: 'lagoobernetes_openshiftRemove',
        branches: 'true',
        pullrequests: 'null',
        openshift: {
          name: 'credentialtest-openshift',
        },
        created: '2018-01-15 11:09:35',
      },
    },
  };

  it('should display error, if GraphQL sends error messages', async () => {
    _castMockForFlow(queryGraphQL).mockImplementationOnce(() =>
      Promise.resolve(mockErrorResponse),
    );

    const clog = jest.fn();
    const cerr = jest.fn();

    const code = await handler({
      clog,
      cerr,
      cwd: 'some/path',
      options: {
        format: 'table',
        token: 'token/path',
        project: 'some_project',
      },
    });

    expect(code).toBe(1);
    expect(cerr.mock.calls).toMatchSnapshot();
  });

  it('should show error on missing project', async () => {
    _castMockForFlow(queryGraphQL).mockImplementationOnce(() =>
      Promise.resolve({}),
    );

    const clog = jest.fn();
    const cerr = jest.fn();

    const code = await handler({
      clog,
      cerr,
      cwd: 'some/path',
      options: {
        format: 'table',
        token: 'token/path',
        project: 'not_existing',
      },
    });

    expect(code).toBe(0);
    expect(clog.mock.calls).toMatchSnapshot();
  });

  it('should show details for given project', async () => {
    _castMockForFlow(queryGraphQL).mockImplementationOnce(() =>
      Promise.resolve(mockResponse1),
    );

    const clog = jest.fn();
    const cerr = jest.fn();

    const code = await handler({
      clog,
      cerr,
      cwd: 'some/path',
      options: {
        format: 'table',
        token: 'token/path',
        project: 'myproject',
      },
    });

    expect(code).toBe(0);
    expect(clog.mock.calls).toMatchSnapshot();
  });
});
