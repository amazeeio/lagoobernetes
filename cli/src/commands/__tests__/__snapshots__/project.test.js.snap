// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handler should display error, if GraphQL sends error messages 1`] = `
Array [
  Array [
    "[31mOops! The Lagoobernetes API returned an error:[39m",
  ],
  Array [
    "[31msomething something error[39m",
  ],
]
`;

exports[`handler should show details for given project 1`] = `
Array [
  Array [
    "[38;2;132;132;132mName[39m                   credentialstest-project1
[38;2;132;132;132mCustomer[39m               credentialtest-customer1
[38;2;132;132;132mGit URL[39m                project1.git
[38;2;132;132;132mActive Systems Deploy[39m  lagoobernetes_openshiftBuildDeploy
[38;2;132;132;132mActive Systems Remove[39m  lagoobernetes_openshiftRemove
[38;2;132;132;132mBranches[39m               true
[38;2;132;132;132mPull Requests[39m          null
[38;2;132;132;132mOpenshift[39m              credentialtest-openshift
[38;2;132;132;132mCreated[39m                2018-01-15 11:09:35",
  ],
]
`;

exports[`handler should show error on missing project 1`] = `
Array [
  Array [
    "No project 'not_existing' found.",
  ],
]
`;
